### ISSUE TITLE
Neotree Update Breaks Workflow

### ISSUE THREAD

**Issue by jupl** [syl20bnr/spacemacs#7446]
#### Description :octocat:

Something changed where Neotree is becoming disruptive to my workflow:
- When I have Neotree open (`SPC p t`) in a project and try to open a terminal using `SPC a s t` the terminal no longer pops up from underneath with the message `split-window: Cannot split side window or parent of side window`. `SPC w p m` to show the message buffer does work though.
- When I have Neotree open (`SPC p t`) in a project and try to open Magit using `SPC g s` with `git-magit-status-fullscreen` enabled the status window appears with Neotree still visible when it wasn't visible before. This is problematic because now when I want to commit (`c c`) it only show one pane instead of two and it doesn't show the `COMMIT_EDITMSG` buffer anymore and have to look for it.

In both cases things work when I hide the Neotree pane, which is an inconvenience to hide every time I want to do something like the above.

**EDIT**: There were some commits to Neotree within a week that may be contributing to this. I did a rollback which had an older Neotree and everything is working.
#### Reproduction guide :beetle:
- Start Emacs with packages up to date
- Follow above steps

_Observed behaviour:_ :eyes: :broken_heart:
- Mentioned above

_Expected behaviour:_ :heart: :smile:
- Not mentioned above
#### System Info :computer:
- OS: darwin
- Emacs: 25.1.1
- Spacemacs: 0.200.2
- Spacemacs branch: master (rev. 7bdde17)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:

``` elisp
(typescript colors docker editorconfig emacs-lisp html javascript markdown org-plus osx react restclient shell shell-scripts syntax-checking vimscript yaml
            (auto-completion :variables auto-completion-return-key-behavior nil auto-completion-tab-key-behavior 'complete auto-completion-enable-help-tooltip t)
            (clojure :variables clojure-enable-fancify-symbols t)
            (git :variables git-magit-status-fullscreen t)
            (version-control :variables version-control-diff-tool 'diff-hl))
```

**Comment by bmag** [comment]
Regarding shell-pop (`SPC a s t`), this seems to be a problem in `shell-pop` when `shell-pop-full-span` is non-nil and neotree is open. @jupl Can you try and reproduce it with stock emacs + neotree + shell-pop, and report upstream to shell-pop?
You can use `SPC q D` to restart emacs with only neotree and shell-pop (without spacemacs), or run `emacs -Q` and evaluate this:
```elisp
(package-initialize)
(require 'neotree)
(require 'shell-pop)
```

Regarding Magit, there's an open PR (which I can't find right now) for using a function provided by Magit, instead of our own custom one. You can try it for yourself by setting `magit-display-buffer-function` to `magit-display-buffer-fullframe-status-v1`, and if that doesn't work then it should be another report upstream (to Magit).

The crucial change in Neotree is that now it displays its buffer in a proper side window (with `display-buffer-in-side-window`). When there is a side window, its parent (the `frame-root-window`) can't be split, unless `ignore-window-parameters` is nil (usually it is let-bound to temporarily ignore parameters). Any feature that uses `(split-window (frame-root-window))` is potentially affected.

**Comment by jupl** [comment]
@bmag Thank you for the thorough explanation!

Regarding `shell-pop` and `neotree` yes I can reproduce the issue on stock Emacs. When I go report it do I just mention `neotree` and `shell-pop-full-span`?

**Comment by bmag** [comment]
You can use the "description", "reproduction guide", "observed behaviour" and "expected behaviour" sections from our bug-report template ([example](https://github.com/syl20bnr/spacemacs/wiki/Debugging#dummy-example-of-a-good-bug-report)). You should definitely mention `neotree` and `shell-pop-full-span`, and maybe also link to my earlier comment.

**Comment by jupl** [comment]
kyagi/shell-pop-el#50

**Comment by jupl** [comment]
I added my environment to a new machine, looks like the issue with `magit` is back again.

**Comment by jupl** [comment]
Did a commit fix this? I just tried develop latest and it appears to be working now.

**Comment by lotuc** [comment]
Encountering the same problem, if I want to show my shell by SPC-' I need to close the neotree first. So pity they cannot show the same time, really annoying.

**Comment by d12frosted** [comment]
Shouldn't it be reported to `neotree` instead of `shell-pop`? Especially after [this comment](https://github.com/kyagi/shell-pop-el/issues/50#issuecomment-256310258)?

**Comment by allquantor** [comment]
Same problem here. Any update on this?

**Comment by vatrat** [comment]
Still same issue here

**Comment by hello-marmalade** [comment]
Any news on this? I'm aware that it seems to be a neotree issue as per d12frosted's comment, but is there anything to be done?

**Comment by d12frosted** [comment]
Someone has to report issue to `neotree` and link it here, so we can track the progress.

**Comment by RomanKapitonov** [comment]
Same issue for me

**Comment by Anntoin** [comment]
A workaround if you're happy with the terminal just in the main window, i.e.

```
+--+--------+
|  |        |
|  |        |
|  |        |
|  +--------+
|  |        |
+--+--------+
```
rather than:
```
 +--+--------+
 |  |        |
 |  |        |
 |  |        |
 +--+--------+
 |           |
 +-----------+
```
is to set `shell-default-full-span nil` (note not shell-pop-full-span)
https://github.com/syl20bnr/spacemacs/tree/master/layers/%2Btools/shell#width-of-the-shell-popup-buffers

**Comment by jaimehrubiks** [comment]
Same issue here. Solution above works pretty good. It even fixes another bug which consists in in-file search SPC s s (swoop) not working properly if neotree is shown.

**Comment by newzubakhin** [comment]
Same issue here.

**Comment by iromise** [comment]
hope this can be fixed soon...

**Comment by Kernael** [comment]
I also encounter the same issue with `helm` on the develop branch.

When using `SPC f f`, using <TAB> to enter a subdirectory or peek at a file will not work and trigger "Cannot split side window or parent of side window" when a Neotree is opened.

**Comment by syl20bnr** [comment]
I have it too, with Treemacs sometimes.

Must close the file tree to be able to open a file with Helm :-(

**Comment by FieryCod** [comment]
Having the same issue. Is there a chance to fix it?

**Comment by thanhvg** [comment]
I think it's a helm's bug. I keep helm and helm core at  "helm-20180527.2241"  and  "helm-core-20180514.500 to avoid this issue

**Comment by skebanga** [comment]
Same happens to me even without Neotree open. I just `SPC f f` and a `HELM find files` mini-buffer opens, but I cannot tab into another directory

**Comment by dentafrice** [comment]
Seeing this as well, closing NeoTree seems to make TAB work again for me.

**Comment by simon-bourne** [comment]
I did a little bit of digging around this issue. The problem, for me at least, is that when Helm tries to get a persistent action window, it sees NeoTree's window is *dedicated* and tries to split it. That fails because NeoTree is a *side window*, and so can't be split. As a very hacky, temporary workaround, I replaced `helm-persistent-action-display-window` in `helm-core-*/helm.el` with:

```elisp
(defun helm-persistent-action-display-window (&optional split-onewindow)
  "Return the window that will be used for persistent action.
If SPLIT-ONEWINDOW is non-`nil' window is split in persistent action."
  (with-helm-window
    (setq helm-persistent-action-display-window
          (cond ((and (window-live-p helm-persistent-action-display-window)
                      (not (member helm-persistent-action-display-window
                                   (get-buffer-window-list helm-buffer))))
                 helm-persistent-action-display-window)
                ((and helm--buffer-in-new-frame-p helm-initial-frame)
                 (with-selected-frame helm-initial-frame (selected-window)))
                (split-onewindow (split-window))
                ((get-mru-window))
                (t split-window (get-mru-window nil t))))))
```

and it all seemed to work. I'll get an issue created in Helm when I get a chance, as I think that's the right place for the fix. In the meantime, it would be great to get some feedback on whether it breaks anything else.

**Comment by dentafrice** [comment]
@simon-bourne this is awesome, thank you <3!  I've added that snippet, wrapped with ` (with-eval-after-load 'helm)` to my .spacemacs and it appears to have resolved the issue for me.

I'll use it for a week or so and let you know if I notice anything broken with Helm as a result.

**Comment by iromise** [comment]
@simon-bourne This works fine in Linux, but in MAC OS it gives the error `window-live-p nil`

**Comment by simon-bourne** [comment]
@iromise Thanks for reporting. There's a couple of things you could try, but I don't have access to a Mac.

Firstly, try this code instead:

```elisp
(defun helm-persistent-action-display-window (&optional split-onewindow)
  "Return the window that will be used for persistent action.
If SPLIT-ONEWINDOW is non-`nil' window is split in persistent action."
  (with-helm-window
      (setq helm-persistent-action-display-window (get-mru-window))))
```

Secondly, let me know what version of emacs you're on.

**Comment by Kernael** [comment]
@simon-bourne Seems to work on MacOS with emacs version 26.1

**Comment by simon-bourne** [comment]
@Kernael The new version I just posted, or the old version I posted a week or so ago? Did you try the old version (and did it work if so). Thanks.

**Comment by Kernael** [comment]
@simon-bourne The new one, but the previous version seems to work as well

**Comment by iromise** [comment]
@simon-bourne  I am sorry it seems that on Linux it also gives the error, my emacs on Linux is
```
GNU Emacs 27.0.50
Copyright (C) 2018 Free Software Foundation, Inc.
GNU Emacs comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of GNU Emacs
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.
```
I have try the code on Ubuntu 18.04 you given just now, still gives the error... Seems strange.

**Comment by simon-bourne** [comment]
Isn't 26.1 the latest released emacs?

**Comment by iromise** [comment]
I compile the Emacs from the source code(in the github).

**Comment by simon-bourne** [comment]
Works fine for me on a checkout from today, using the develop branch of spacemacs (open emacs, open recent project `SPC pp`, open neotree `SPC pt`, find project files `SPC pf`). I'm on Linux Mint 19, which is based off Ubuntu 18.04. Things to try:
- Maybe it's your config. Try a minimal .spacemacs with just this fix.
- Try a build from todays emacs source

**Comment by bmag** [comment]
TL;DR: the bug only affects Emacs 25 and master branch (together), although Emacs 25 users on develop branch are protected only because of a bug in window-purpose. Rough draft of a solution below.

Neotree is displayed in a [side window](https://www.gnu.org/software/emacs/manual/html_node/elisp/Side-Windows.html#Side-Windows).

In Emacs 25 and older: Normally, the frame's root window can't be split when a side window exists, and also a side window itself can't be split. Trying either results in error: `Cannot split side window or parent of side window`.

In Emacs 26 and newer: The frame's root *can* be split when a side window exists, but splitting a side window still throws the same error (normally).

In Spacemacs develop branch: by default, `purpose-mode` is enabled, which overrides how Neotree buffer is displayed, technically making it not a side window anymore. (it's actually a bug in window-purpose, the side window property shouldn't be lost)

In summary, if you have Emacs 26.1 or later, or if you're on develop branch (and didn't disable window-purpose), then you aren't affected by the bugs above. Otherwise, you are.

We can fix this by customizing how Neotree buffer is displayed. It won't change the situation for develop branch users (unless the bug in window-purpose is fixed), but the behavior should be close enough to what develop branch has already:

```elisp
(defun spacemacs/display-neotree (buffer _alist)
  "Display BUFFER to the left or right of the main window.
Doesn't give the new window real side window parameters, so
technically it is not a side window.

_ALIST is ignored."
  (let ((side (if (eq neo-window-position 'left) 'left 'right))
        (main-window (if (version< emacs-version "26")
                         (window--major-non-side-window)
                       (window-main-window))))

    ;; don't set width or dedicate buffer, even though we could, because
    ;; neotree will do that later
    (window--display-buffer buffer
                            (split-window main-window nil side)
                            'window)))

(setq neo-display-action '(spacemacs/display-neotree))
```

**Comment by iromise** [comment]
@simon-bourne I am sorry, I forget I am use the treemacs. But it really gives me the error when I AM USING THE ABOVE CODE with Ubuntu 18.04 and emacs 27.0.50.

**Comment by simon-bourne** [comment]
@iromise Ok, sounds like it's worth investigating further.

@bmag Thanks for chiming in - I'm not that familiar with the code base or elisp yet, so it's appeciated! That also explains why I can delete everything except the neotree window in develop on 26.1. One thing to note is that I was slightly wrong in my initial analysis. Helm is trying to split it's **own** window (also a side window in spacemacs) when it sees that `next-window` (neotree) is dedicated. It seems opening a side window on the main editor window (for helm) causes the window list to go from [editor window, neotree] to [helm, neotree, editor window].

I still see problems when neotree is displayed with emacs 26.1 and develop branch (caused by helm attempting to split it's window). For example, try tab completing when renaming from neotree, or tab completing in find file (`SPC ff`).

My plan was to ask for a new configuration point in `helm-persistent-action-display` (or whatever it's called), where we can call something like `get-mru-window`. I think generally, splitting the helm window might be a good strategy for some helm users, but it's never what we want to do (as we open helm in a side window).

**Comment by simon-bourne** [comment]
@iromise This still works for me. Here's the exact steps I'm trying. Can you try them as well, on the `develop` branch (obviously be careful to save anything you want to keep):

```bash
cd ~/.emacs.d
git checkout develop
cd
git clone https://github.com/emacs-mirror/emacs
cd emacs
make -j 8
mv ~/.spacemacs ~/spacemacs.orig
src/emacs
```
- Select vim editing style and standard distro
- `SPC fed` to open your newly created .spacemacs and make your user config function look like:
```elisp
(defun dotspacemacs/user-config ()
  "Configuration for user code:
This function is called at the very end of Spacemacs startup, after layer
configuration.
Put your configuration code here, except for variables that should be set
before packages are loaded."
  (with-eval-after-load "helm"
      (defun helm-persistent-action-display-window (&optional split-onewindow)
        "Return the window that will be used for persistent action.
If SPLIT-ONEWINDOW is non-`nil' window is split in persistent action."
        (with-helm-window
         (setq helm-persistent-action-display-window (get-mru-window)))))
  )
```
Save and restart emacs.
`SPC 0` to open treemacs
`SPC ff` to open find files and try tab completion. Works perfectly for me.

**Comment by iromise** [comment]
I will try it as soon as I go back to work(But it will take several days). But I remember that although it gives the error, I still can use C-x-C-f to find file, and use `->` to go into children folder.

Test on Ubuntu 18.04

**Comment by iromise** [comment]
For now, I test it on mac with 
```
        (defun helm-persistent-action-display-window (&optional split-onewindow)
          "Return the window that will be used for persistent action.
If SPLIT-ONEWINDOW is non-`nil' window is split in persistent action."
          (with-helm-window
            (setq helm-persistent-action-display-window
                  (cond ((and (window-live-p helm-persistent-action-display-window)
                              (not (member helm-persistent-action-display-window
                                           (get-buffer-window-list helm-buffer))))
                         helm-persistent-action-display-window)
                        ((and helm--buffer-in-new-frame-p helm-initial-frame)
                         (with-selected-frame helm-initial-frame (selected-window)))
                        (split-onewindow (split-window))
                        ((get-mru-window))
                        (t split-window (get-mru-window nil t))))))
        (with-eval-after-load 'helm (helm-persistent-action-display-window))
```
And this also gives the error
```
(Spacemacs) Error in dotspacemacs/user-config: Wrong type argument: window-live-p, nil
```
The Emacs on MacOS is 
```
➜  .emacs.d git:(develop) emacs --version
GNU Emacs 26.1
Copyright (C) 2018 Free Software Foundation, Inc.
GNU Emacs comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of GNU Emacs
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.
```

**Comment by simon-bourne** [comment]
Your with-eval-after-load is all wrong. Just use this one:

```elisp
  (with-eval-after-load "helm"
      (defun helm-persistent-action-display-window (&optional split-onewindow)
        "Return the window that will be used for persistent action.
If SPLIT-ONEWINDOW is non-`nil' window is split in persistent action."
        (with-helm-window
         (setq helm-persistent-action-display-window (get-mru-window)))))
```

**Comment by simon-bourne** [comment]
It's slowly dawning on me that there's 2 bugs here, both manifesting with the same error and both when neotree is visible. There's one @bmag mentions, related to not being able to split the root window when neotree is visible as a side window. There's another bug being discussed when tab completing in helm, related to not being able to split the helm window. I'll get a new issue and pull request created for the latter when I get a chance, but just wanted to clear up any confusion for now.

**Comment by iromise** [comment]
Sorry for my wrong code, it works when using latter code snippet.

**Comment by ArenT1981** [comment]
@simon-bourne Your fix worked great for me, thanks!

Incidentally, I actually encountered this issue (which subsequently lead me here) with minimap-mode enabled, so it was the same underlying bug/issue with helm. So thought I'd drop a comment here in case anyone else's search for minimap and helm tab-completion leads them to this page...

**Comment by duianto** [comment]
The `cannot split side window` message when completing a directory in helm find files `SPC f f` with Neotree open, seems to have been solved by `sdwolfz`.

The fix has been merged upstream.
Fix splitting issue when invoking helm from a dedicated window https://github.com/emacs-helm/helm/pull/2200

<details><summary>
System info
</summary>
<pre>
#### System Info :computer:
- OS: windows-nt
- Emacs: 26.3
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 7dd411882)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(autohotkey auto-completion
            (dart :variables dart-server-sdk-path "c:/src/flutter/bin/cache/dart-sdk/" dart-server-enable-analysis-server t)
            emacs-lisp ess git helm html
            (lsp :variables lsp-clients-dart-server-command "~/AppData/Roaming/Pub/Cache/bin/dart_language_server" lsp-dart-analysis-sdk-dir "c:/src/flutter/bin/cache/dart-sdk/")
            multiple-cursors org php shell spell-checking syntax-checking themes-megapack neotree typescript version-control vinegar)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG SOUND NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS THREADS LCMS2
</pre>
</details>

**Comment by vv111y** [comment]
`SPC f f` problem started happening to me a couple weeks ago, the error is 
```
helm-select-persistent-action-window: Wrong number of arguments: (lambda (&optional split-onewindow) "Return the window that will be used for persistent action.
If SPLIT-ONEWINDOW is non-`nil' window is split in persistent action." (with-helm-window (setq helm-persistent-action-display-window (get-mru-window)))), 2
```
Everything is updated, so it seems https://github.com/emacs-helm/helm/pull/2200 does not solve whatever is happening to me.

**Comment by duianto** [comment]
@vv111y Do you have any reproduction steps?

It's not the same message as your seeing but there is another `cannot split window ...` issue if a help buffer is open.

There's more info in this comment: https://github.com/syl20bnr/spacemacs/issues/11368#issuecomment-532080370

**Comment by vv111y** [comment]
Let me know if this isn't what you are looking for:
- the problem is there right away. Simply `SPC f f` 
- I can back up to parent directories, but trying to go into a child directory causes the error

Sorry I didn't keep track, so I can't say which commit did it, whether it was from spacemacs, or from helm update. Started about a month ago.

**Comment by duianto** [comment]
Are you still using the `helm-persistent-action-display-window` workaround from the comment above? https://github.com/syl20bnr/spacemacs/issues/7446#issuecomment-417376425

Try commenting it out, it was about a month ago that the cannot split window issue from Neotree or Treemacs was fixed upstream in Helm:
Fix splitting issue when invoking helm from a dedicated window emacs-helm/helm#2200

Maybe the workaround now is causing issues.

**Comment by vv111y** [comment]
I feel silly, yes that was still in my config. I forgot about it. Thanks

**Comment by duianto** [comment]
There's no need to feel silly, the hole was fixed (upstream) under the hole covering (workaround) so it was hard to see it 😄
